### PR TITLE
Add Reaper

### PR DESCRIPTION
**Pull request by CleoMenezesJr**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

File: fm.reaper.Reaper.desktop [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[Desktop Entry]
+Name=Reaper
+Exec=/app/REAPER/reaper
```
Comment (hfiguiere) [inline fm.reaper.Reaper.desktop:3]: you instead have a script or a symlink to this in `/app/bin` and just put the name of that. Not an absolute path to a non standard location.

File: fm.reaper.Reaper.metainfo.xml [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  <summary>Digital Audio Workstation.</summary>
+  <url type="homepage">https://www.reaper.fm/</url>
+  <url type="bugtracker">https://forum.cockos.com/forumdisplay.php?f=22</url>
+  <url type="help">https://forum.cockos.com/</url>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>LicenseRef-proprietary=https://www.reaper.fm/dist-agreement.php</project_license>
+  <description>
+    <p>REAPER is a complete digital audio production application for computers, offering a full multitrack audio and MIDI recording, editing, processing, mixing and mastering toolset.  </p>
+    <p> Using REAPER you can import any audio and MIDI, synthesize, sample, compose, arrange, edit, mix, and master songs or any other audio projects. If you add a hardware audio interface and a microphone, you have a complete professional quality recording studio, suitable for recording anything from a soloist to a band to an orchestra.</p>
+    <p>"REAPER" is a trial shareware version.</p>
```
Comment (hfiguiere) [inline fm.reaper.Reaper.metainfo.xml:14]: usually there is a disclaimer that this is unofficial.

File: fm.reaper.Reaper.yml [hunk 3]
```diff
@@ -0,0 +1,70 @@
+app-id: fm.reaper.Reaper
+runtime: org.freedesktop.Platform
+runtime-version: '21.08'
+sdk: org.freedesktop.Sdk
+tags:
+ - proprietary
+command: /app/REAPER/reaper
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:7]: command is not an absolute path either. See my comment on the `.desktop`

File: fm.reaper.Reaper.yml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  - /include
+  - /cache
+  - /include
+
+add-extensions:
+  org.freedesktop.LinuxAudio.Plugins:
+    directory: extensions/Plugins
+    version: '21.08'
+    add-ld-path: lib
+    merge-dirs: ladspa;dssi;lv2;lxvst;vst3
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:38]:
```suggestion
    merge-dirs: ladspa;dssi;lv2;vst;vst3
```

File: fm.reaper.Reaper.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  - --filesystem=xdg-run/pipewire-0
+  - --device=all
+  - --filesystem=home
+  - --system-talk-name=org.freedesktop.RealtimeKit1
+  - --system-talk-name=org.freedesktop.Flatpak
+  - --env=ALSA_CONFIG_PATH=
+  - --env=LV2_PATH=/app/extensions/Plugins/lv2
+  - --env=DSSI_PATH=/app/extensions/Plugins/dssi
+  - --env=LADSPA_PATH=/app/extensions/Plugins/ladspa
+  - --env=VST_PATH=/app/extensions/Plugins/lxvst
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:23]:
```suggestion
  - --env=VST_PATH=/app/extensions/Plugins/vst
```

File: fm.reaper.Reaper.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  - --socket=pulseaudio
+  - --filesystem=xdg-run/pipewire-0
+  - --device=all
+  - --filesystem=home
+  - --system-talk-name=org.freedesktop.RealtimeKit1
+  - --system-talk-name=org.freedesktop.Flatpak
+  - --env=ALSA_CONFIG_PATH=
+  - --env=LV2_PATH=/app/extensions/Plugins/lv2
+  - --env=DSSI_PATH=/app/extensions/Plugins/dssi
+  - --env=LADSPA_PATH=/app/extensions/Plugins/ladspa
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:22]: does Reaper really support these plugins formats?

File: fm.reaper.Reaper.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  - --env=VST_PATH=/app/extensions/Plugins/lxvst
+  - --env=VST3_PATH=$HOME/.vst3:/app/extensions/Plugins/vst3
+
+cleanup:
+  - '*.a'
+  - '*.la'
+  - /include
+  - /cache
+  - /include
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:31]: I doubt there is any of these to cleanup. Also duplicate entry.

File: fm.reaper.Reaper.yml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+modules:
+  - name: reaper
+    buildsystem: simple
+    build-commands:
+      - sh install-reaper.sh --quiet --install /app
+      - install -Dm644 /app/REAPER/Resources/main.png /app/share/icons/hicolor/256x256/apps/fm.reaper.Reaper.png 
+      - install -Dm644 fm.reaper.Reaper.desktop /app/share/applications/${FLATPAK_ID}.desktop
+      - install -Dm644 $FLATPAK_ID.metainfo.xml /app/share/metainfo/$FLATPAK_ID.metainfo.xml
+      - desktop-file-validate /app/share/applications/${FLATPAK_ID}.desktop
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:60]:
```suggestion
```

File: fm.reaper.Reaper.yml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    sources:
+      - type: archive
+        url: https://www.reaper.fm/files/6.x/reaper658_linux_x86_64.tar.xz
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:63]: since this is x86_64 only, you need to tell Flathub to not build anything else.

See https://github.com/flathub/flathub/wiki/App-Maintenance#only-build-on-x86_64

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: fm.reaper.Reaper.yml [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+finish-args:
+  - --share=ipc
+  - --share=network
+  - --socket=x11
+  - --socket=pulseaudio
+  - --filesystem=xdg-run/pipewire-0
+  - --device=all
+  - --filesystem=home
+  - --system-talk-name=org.freedesktop.RealtimeKit1
+  - --system-talk-name=org.freedesktop.Flatpak
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:18]: do you really need this ?

File: fm.reaper.Reaper.desktop (hunk 1 above)
Comment (CleoMenezesJr) [reply fm.reaper.Reaper.desktop:3]: When I put only the name in it I get the error: `WARNING: Binary not found for Exec line in /var/home/cleomenezesjr/Projects/Reaper/build/export/share/applications/fm.reaper.Reaper.desktop: reaper`

Comment (hfiguiere) [reply fm.reaper.Reaper.desktop:3]: as I said, have a symlink in `/app/bin`. Or wrapper script.

File: fm.reaper.Reaper.yml (hunk 6 above)
Comment (CleoMenezesJr) [reply fm.reaper.Reaper.yml:22]: I wanted to leave the option of the types of plugins that the user could use. Too much?

File: fm.reaper.Reaper.yml (hunk 10 above)
Comment (CleoMenezesJr) [reply fm.reaper.Reaper.yml:18]: ` - --device=all`
Wouldn't it be necessary to be able to recognize any Audio Interface? I also put because Reaper can edit video.
`- --filesystem=home`
As the project and plugins could be anywhere on the home, I decided to give access to it.
```
  - --system-talk-name=org.freedesktop.RealtimeKit1
  - --system-talk-name=org.freedesktop.Flatpak
```
Those really should have been removed. Only Realtime that I was confused if it would be necessary for the video editing.

File: fm.reaper.Reaper.yml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  - --env=ALSA_CONFIG_PATH=
+  - --env=LV2_PATH=/app/extensions/Plugins/lv2
+  - --env=DSSI_PATH=/app/extensions/Plugins/dssi
+  - --env=LADSPA_PATH=/app/extensions/Plugins/ladspa
+  - --env=VST_PATH=/app/extensions/Plugins/vst
+  - --env=VST3_PATH=$HOME/.vst3:/app/extensions/Plugins/vst3
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:24]: note: this won't work. $HOME is not substituted. The trick is to have a wrapper script and set it there. Note that in theory a VST3 host should always look into `$HOME/.vst3` so I'm not sure if this is necessary.

 LV2 need the same treatment.

File: fm.reaper.Reaper.yml (hunk 6 above)
Comment (hfiguiere) [reply fm.reaper.Reaper.yml:22]: what option? Does Reaper support them?

File: fm.reaper.Reaper.yml (hunk 10 above)
Comment (hfiguiere) [reply fm.reaper.Reaper.yml:18]: I was only questioning the last line. 

As for `org.freedesktop.RealtimeKit1` it depends.

File: fm.reaper.Reaper.yml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+tags:
+ - proprietary
+command: /app/REAPER/reaper
+
+finish-args:
+  - --share=ipc
+  - --share=network
+  - --socket=x11
+  - --socket=pulseaudio
+  - --filesystem=xdg-run/pipewire-0
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:14]: Does Reaper use JACK?

File: fm.reaper.Reaper.yml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    no-autodownload: true
+
+  org.freedesktop.Platform.ffmpeg-full:
+    directory: lib/ffmpeg
+    version: '21.08'
+    add-ld-path: "."
+    autodelete: false
+
+cleanup-commands:
+- mkdir -p /app/lib/ffmpeg
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:42]: I'd just put this in the same place the creation of `/app/extensions`.

File: fm.reaper.Reaper.yml (hunk 12 above)
Comment (CleoMenezesJr) [reply fm.reaper.Reaper.yml:14]: Yes it can.

Comment (CleoMenezesJr) [reply fm.reaper.Reaper.yml:14]: Yes, it can use.
image

File: fm.reaper.Reaper.yml (hunk 13 above)
Comment (hfiguiere) [reply fm.reaper.Reaper.yml:42]: I didn't review these changes

lot are just copy / paste cargo cult.

File: fm.reaper.Reaper.metainfo.xml [hunk 14]
```diff
@@ -0,0 +1,141 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop-application">
+  <id>rm.reaper.Reaper</id>
```
Comment (hfiguiere) [inline fm.reaper.Reaper.metainfo.xml:3]:
```suggestion
  <id>fm.reaper.Reaper</id>
```

Freudian slip?

(that's why the build fails)

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

File: fm.reaper.Reaper.yml [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+      - type: archive
+        url: https://www.reaper.fm/files/6.x/reaper658_linux_x86_64.tar.xz
+        sha256: 8d18e110a75f99640161c852acf3610dfeb2a2712a49df2012888236e89dcac3
+      - type: file
+        path: fm.reaper.Reaper.desktop
+      - type: file
+        path: fm.reaper.Reaper.metainfo.xml
+    post-install:
+      - install -d /app/extensions/Plugins
+      - mkdir -p /app/lib/ffmpeg
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:55]:
```suggestion
      - install -d /app/lib/ffmpeg
```

just to be consistent.

File: fm.reaper.Reaper.yml [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+finish-args:
+  - --share=ipc
+  - --share=network
+  - --socket=x11
+  - --socket=pulseaudio
+  - --filesystem=xdg-run/pipewire-0
+  - --device=all
+  - --filesystem=home
+  - --system-talk-name=org.freedesktop.RealtimeKit1
+  - --env=ALSA_CONFIG_PATH=
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:18]: so you removed the `_PATH` for the various plugins. But they aren't set anywhere else.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: fm.reaper.Reaper.yml (hunk 16 above)
Comment (CleoMenezesJr) [reply fm.reaper.Reaper.yml:18]: What would be the best way to do this?

**Comment by CleoMenezesJr** [comment]
Aside from here things started to get a little hazy for me. I can't find anything in the documentation that can help me.

**Comment by barthalion** [comment]
bot, build fm.reaper.Reaper

**Comment by barthalion** [comment]
It looks good to me, but I know nothing about audio applications on Linux. Is it good to be merged @hfiguiere @CleoMenezesJr?

**Comment by hfiguiere** [comment]
I can't tell if plugins will work or not. I suggested some things, and noted the things I knew wouldn't work (like `$HOME` not being substituted) - as it's not specific to the application.

I'm not stopping anyone from approving it.

**Comment by CleoMenezesJr** [comment]
> I can't tell if plugins will work or not. I suggested some things, and noted the things I knew wouldn't work (like `$HOME` not being substituted) - as it's not specific to the application.
> 
> I'm not stopping anyone from approving it.

I really don't get the point, plugins by default are in .~/vst and/or ~/.vst3, plugins are scanned by default there.

**Comment by barthalion** [comment]
My understanding is that if you don't follow what other apps using LinuxAudio plugins do, they simply won't work. The default `~/.vst` path is not going to work with flatpaked plugins.

**Comment by Be-ing** [comment]
From REAPER/EULA.txt in https://www.reaper.fm/files/6.x/reaper658_linux_x86_64.tar.xz :

> 3.2  You may not sell, rent, lease, sublicense, transfer, resell for profit or otherwise distribute the Software, its documentation, or any part thereof.

**Comment by hfiguiere** [comment]
That mean https://github.com/flathub/flathub/wiki/App-Requirements#license

File: flathub.json [hunk 17]
```diff
@@ -0,0 +1,3 @@
+{
+  "only-arches": ["x86_64"]
```
Comment (hfiguiere) [inline flathub.json:2]: about this:

image


tl;dr please add aarch64 support.

Comment (hfiguiere) [reply flathub.json:2]: and if this contradicts https://github.com/flathub/flathub/pull/3222/files#r884392645 it's because I assumed you had checked.

Comment (eshimoniak) [reply flathub.json:2]: @hfiguiere I added a simple update for multiarch support [here](https://github.com/CleoMenezesJr/flathub/pull/2), but I don't have an aarch64 machine to test it on. I'll try it in a virtual machine when I get the chance, but it would be nice to have someone verify that this works (or why it doesn't) on real hardware).

Comment (hfiguiere) [reply flathub.json:2]: then why is it not here? the best way to test is to start by building it.

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

File: flathub.json (hunk 17 above)
Comment (CleoMenezesJr) [reply flathub.json]: Now it's here, now wait for the build

Comment (CleoMenezesJr) [reply flathub.json]: @hfiguiere It really was something that went unnoticed for me. My bad

**Comment by eshimoniak** [comment]
It looks like the problem is just the formatting for the changelogs in the release description (too many `<p>` elements). Would changing them to `<span>` elements help?

**Comment by eshimoniak** [comment]
I couldn't figure out how to bulk change all of the `<p>` elements in the "suggest changes" interface, so I had to do it in a PR https://github.com/CleoMenezesJr/flathub/pull/3

**Comment by Be-ing** [comment]
So you're going to blatantly ignore the license??

**Comment by eshimoniak** [comment]
@Be-ing The checklist in the PR template requires permission from the upstream developer in order to merge (as far as I can tell). REAPER does have a formal process to allow for redistribution [detailed on their website](https://www.reaper.fm/dist-agreement.php), but I'm not sure who would need to sign off on it; presumably someone affiliated with flathub.

**Comment by hfiguiere** [comment]
Why not using `extra-data` like it is done for other non redistributable packages?

**Comment by Be-ing** [comment]
How would that change anything legally? Isn't that still distributing it?

**Comment by hfiguiere** [comment]
No. `extra-data` mean that when the user install the flatpak they download the extra-data and it gets installed at that point. **It's just automated downloading.**  

That's what is used for a lot of other proprietary apps.

**Comment by CleoMenezesJr** [comment]
> Why not using `extra-data` like it is done for other non redistributable packages?

Great! Im gonna do it!

**Comment by CleoMenezesJr** [comment]
`type: extra-data` fails to build by apparently not downloading the files.
```yml
    sources:
      - type: extra-data
        filename: reaper_linux_x86_64.tar.xz
        url: https://www.reaper.fm/files/6.x/reaper661_linux_x86_64.tar.xz
        sha256: 2e9d84564cb5fabb9cf3a2dd326be16488aca189743b176456737787fbb4f484
        size: 11675288
        only-arches:
          - x86_64
      - type: extra-data
        filename: reaper_linux_aarch64.tar.xz
        url: https://www.reaper.fm/files/6.x/reaper661_linux_aarch64.tar.xz
        sha256: b0f2c2300fffc407565e363bfbe47a78b592e7a662d2efbd8eeec9948086bd57
        size: 10463064
        only-arches:
          - aarch64
```

```output
❯ flatpak-builder --user --install build --force-clean fm.reaper.Reaper.yml
Emptying app dir 'build'
Downloading sources
Starting build of fm.reaper.Reaper
Cache miss, checking out last cache hit
========================================================================
Building module reaper in /var/home/cleomenezesjr/Projects/flathub/.flatpak-builder/build/reaper-7
========================================================================
Running: sh install-reaper.sh --quiet --install /app
sh: install-reaper.sh: No such file or directory
Error: module reaper: Child process exited with code 127
```

Any suggestion what it could be? With `type: archive` works without any problem.

File: fm.reaper.Reaper.metainfo.xml [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+  <screenshots>
+    <screenshot type="default">
+      <caption>Overview</caption>
+      <image>https://www.reaper.fm/v6img/ss_persp_v63.jpg</image>
+    </screenshot>
+  </screenshots>
+  <releases>
+    <release version="v6.61" date="2022-06-17">
+      <description>
+        <span>ReaSamploMatic5000</span>
```
Comment (JakobDev) [inline fm.reaper.Reaper.metainfo.xml:30]: According to the [Appstream specification](https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-description) `<span>` is not allowed. appstreamcli, which is the more modern reference implementation shows the validation error:
```
$ appstreamcli validate fm.reaper.Reaper.metainfo.xml
E: fm.reaper.Reaper:229: description-markup-invalid span
E: fm.reaper.Reaper:80: description-markup-invalid span
E: fm.reaper.Reaper:95: description-markup-invalid span
E: fm.reaper.Reaper:43: description-markup-invalid span
E: fm.reaper.Reaper:193: description-markup-invalid span
I: fm.reaper.Reaper:5: summary-has-dot-suffix Digital Audio Workstation.
E: fm.reaper.Reaper:223: description-markup-invalid span
E: fm.reaper.Reaper:105: description-markup-invalid span
E: fm.reaper.Reaper:30: description-markup-invalid span
E: fm.reaper.Reaper:187: description-markup-invalid span
E: fm.reaper.Reaper:120: description-markup-invalid span
E: fm.reaper.Reaper:75: description-markup-invalid span
E: fm.reaper.Reaper:150: description-markup-invalid span
E: fm.reaper.Reaper:217: description-markup-invalid span
E: fm.reaper.Reaper:90: description-markup-invalid span
E: fm.reaper.Reaper:181: description-markup-invalid span
E: fm.reaper.Reaper:115: description-markup-invalid span
E: fm.reaper.Reaper:100: description-markup-invalid span
E: fm.reaper.Reaper:211: description-markup-invalid span
E: fm.reaper.Reaper:130: description-markup-invalid span
E: fm.reaper.Reaper:175: description-markup-invalid span
E: fm.reaper.Reaper:138: description-markup-invalid span
E: fm.reaper.Reaper:70: description-markup-invalid span
E: fm.reaper.Reaper:85: description-markup-invalid span
E: fm.reaper.Reaper:63: description-markup-invalid span
E: fm.reaper.Reaper:205: description-markup-invalid span
E: fm.reaper.Reaper:56: description-markup-invalid span
E: fm.reaper.Reaper:49: description-markup-invalid span
E: fm.reaper.Reaper:235: description-markup-invalid span
E: fm.reaper.Reaper:161: description-markup-invalid span
E: fm.reaper.Reaper:169: description-markup-invalid span
E: fm.reaper.Reaper:199: description-markup-invalid span
E: fm.reaper.Reaper:110: description-markup-invalid span
E: fm.reaper.Reaper:125: description-markup-invalid span

Validation failed: errors: 33, infos: 1
```

**Comment by hfiguiere** [comment]
that's because everything is installed with the "apply_extra" script that needs to be provided.

The build will just build an empty flatpak and in installation (or update) it will download the extra-data and run the apply_extra script to "install" it.

Many examples
https://github.com/search?q=org%3Aflathub+apply_extra&type=code

**Comment by CleoMenezesJr** [comment]
> that's because everything is installed with the "apply_extra" script that needs to be provided.
> 
> The build will just build an empty flatpak and in installation (or update) it will download the extra-data and run the apply_extra script to "install" it.
> 
> Many examples https://github.com/search?q=org%3Aflathub+apply_extra&type=code
I saw several examples and tested it in some ways, but I can't understand why it doesn't download the `source`.

```yml
modules:
  - name: reaper
    buildsystem: simple
    build-commands:
      - mkdir -p ${FLATPAK_DEST}/bin
      - install -D apply_extra ${FLATPAK_DEST}/bin/apply_extra
      - sh install-reaper.sh --quiet --install /app
      - install -Dm644 /app/REAPER/Resources/main.png /app/share/icons/hicolor/256x256/apps/fm.reaper.Reaper.png
      - install -Dm644 fm.reaper.Reaper.desktop /app/share/applications/${FLATPAK_ID}.desktop
      - install -Dm644 $FLATPAK_ID.metainfo.xml /app/share/metainfo/$FLATPAK_ID.metainfo.xml
      - mkdir -p /app/bin
      - ln -sf /app/REAPER/reaper /app/bin/reaper
    sources:
      - type: extra-data
        filename: reaper_linux_x86_64.tar.xz
        url: https://www.reaper.fm/files/6.x/reaper661_linux_x86_64.tar.xz
        sha256: 2e9d84564cb5fabb9cf3a2dd326be16488aca189743b176456737787fbb4f484
        size: 11675288
        only-arches:
          - x86_64
      - type: script
        dest-filename: apply_extra
        commands:
          - tar xf reaper_linux_x86_64.tar.xz
          - rm -f reaper_linux_x86_64.tar.xz

      - type: extra-data
        filename: reaper_linux_aarch64.tar.xz
        url: https://www.reaper.fm/files/6.x/reaper661_linux_aarch64.tar.xz
        sha256: b0f2c2300fffc407565e363bfbe47a78b592e7a662d2efbd8eeec9948086bd57
        size: 10463064
        only-arches:
          - aarch64
      - type: script
        dest-filename: apply_extra
        commands:
          - tar xf reaper_linux_aarch64.tar.xz
          - rm -f reaper_linux_aarch64.tar.xz

      - type: file
        path: fm.reaper.Reaper.desktop
      - type: file
        path: fm.reaper.Reaper.metainfo.xml

    post-install:
      - install -d /app/extensions/Plugins
      - install -d /app/lib/ffmpeg

```

Output:
```output
❯ flatpak-builder --user --install build --force-clean fm.reaper.Reaper.yml
Downloading sources
Initializing build dir
Committing stage init to cache
Starting build of fm.reaper.Reaper
========================================================================
Building module reaper in /var/home/cleomenezesjr/Projects/flathub/.flatpak-builder/build/reaper-1
========================================================================
Running: mkdir -p ${FLATPAK_DEST}/bin
Running: install -D apply_extra ${FLATPAK_DEST}/bin/apply_extra
Running: sh install-reaper.sh --quiet --install /app
sh: install-reaper.sh: No such file or directory
Error: module reaper: Child process exited with code 127
```

**Comment by hfiguiere** [comment]
because at the time of the "build" the archive isn't downloaded. Most of your build commands should be run in the apply_extra.

**Comment by barthalion** [comment]
@CleoMenezesJr Please look into other apps using extra-data like Spotify and Discord for examples.

**Comment by eshimoniak** [comment]
[Discord is also using `archive` instead of `extra-data`](https://github.com/flathub/com.discordapp.Discord/blob/87163dcfa3df0b70550e977a1bf601608ec23e0b/com.discordapp.Discord.json#L113). Looking at Spotify (and an assortment of other apps using `apply_extra`) seems to suggest that all the `apply_extra` script needs to do is extract the archive and move it where it needs to go.

Running with `--build-only` does not throw any errors. However, running with `--finish-only` gives `Error: Command 'reaper' not found`. It seems like the problem is simple enough, but I can't find the cause.

Here's my current draft manifest: https://github.com/eshimoniak/flathub/blob/extra-data/fm.reaper.Reaper.yml, any help is appreciated.

File: flathub.json [hunk 19]
```diff
@@ -0,0 +1,3 @@
+{
+  "only-arches": ["x86_64", "aarch64"]
```
Comment (hfiguiere) [inline flathub.json:2]: you don't need this file. These are the only two architectures.

**Comment by eshimoniak** [comment]
@barthalion Thanks! I didn't realize that there needed to be a wrapper script to satisfy the build condition for an executable. [I've openned a PR](https://github.com/CleoMenezesJr/flathub/pull/4) against @CleoMenezesJr's branch because a big update to the metainfo file was also required to satisfy the validator.

**Comment by CleoMenezesJr** [comment]
bot, build

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

File: flathub.json (hunk 19 above)
Comment (hfiguiere) [reply flathub.json:2]: still relevant.

File: fm.reaper.Reaper.yml [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+
+finish-args:
+  - --share=ipc
+  - --share=network
+  - --socket=x11
+  - --socket=pulseaudio
+  - --filesystem=xdg-run/pipewire-0
+  - --device=all
+  - --filesystem=home
+  - --system-talk-name=org.freedesktop.RealtimeKit1
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:17]:
```suggestion
```

File: fm.reaper.Reaper.yml [hunk 21]
```diff
@@ -0,0 +1,83 @@
+app-id: fm.reaper.Reaper
+runtime: org.freedesktop.Platform
+runtime-version: '21.08'
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:3]: 22.08 is out.

**Review by hfiguiere — CHANGES_REQUESTED** [review]

**Comment by Skaldebane** [comment]
Any updates here?

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

File: fm.reaper.Reaper.desktop [hunk 22]
```diff
@@ -0,0 +1,9 @@
+[Desktop Entry]
+Name=Reaper
+Exec=reaper
+# Icon is validated at build time, but it doesn't extist yet because apply_extra
+# hasn't run yet.
+# Icon=/app/extra/REAPER/Resources/main.png
```
Comment (hfiguiere) [inline fm.reaper.Reaper.desktop:6]: you need to actually add an icon here too.

File: fm.reaper.Reaper.yml [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+        only-arches: [aarch64]
+      - type: script
+        dest-filename: reaper
+        commands:
+          - /app/extra/REAPER/reaper
+      - type: file
+        path: fm.reaper.Reaper.desktop
+      - type: file
+        path: fm.reaper.Reaper.metainfo.xml
+    post-install:
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:80]: you don't need a `post-install` with buildsystem `simple`. All goes into the `build-commands`

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

File: fm.reaper.Reaper.desktop (hunk 22 above)
Comment (CleoMenezesJr) [reply fm.reaper.Reaper.desktop]: @hfiguiere For some reason, I can't place the icon for apparently some lack of permission (?).
`install: cannot create directory '/app/share/icons': Read-only file system`

Comment (CleoMenezesJr) [reply fm.reaper.Reaper.desktop]: Change this apparently fixed it:
```
-           - install -Dm644 /app/extra/REAPER/Resources/main.png /app/share/icons/hicolor/${size}x${size}/apps/${FLATPAK_ID}.png
+           - install -Dm644 /app/extra/REAPER/Resources/main.png $FLATPAK_APP_DIR/share/icons/hicolor/${size}x${size}/apps/${FLATPAK_ID}.png

```

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

File: fm.reaper.Reaper.desktop (hunk 22 above)
Comment (CleoMenezesJr) [reply fm.reaper.Reaper.desktop]: But still can't find the icon. :/

File: fm.reaper.Reaper.yml [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+        commands:
+          # Extract downnloaded tarball
+          - mkdir -p export/share
+          - tar xf reaper.tar.xz -C /app/extra/export/share
+          - rm -f reaper.tar.xz
+          # Run installer script
+          - cd export/share/reaper_linux_*
+          - ./install-reaper.sh --quiet --install /app/extra
+          # Save icon
+          - install -Dm644 /app/extra/REAPER/Resources/main.png $FLATPAK_APP_DIR/share/icons/hicolor/${size}x${size}/apps/${FLATPAK_ID}.png
```
Comment (hfiguiere) [inline fm.reaper.Reaper.yml:60]: as I said, the icon needs to be installed by the package, not the extra data. because at build time the extra data doesn't exists. Not sure about which icon to use though.

Comment (CleoMenezesJr) [reply fm.reaper.Reaper.yml:60]: Thanks. Now it's working nicely.

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

**Comment by Skaldebane** [comment]
Great! Didn't expect my message to start this chain of events so quickly, but anyways, i'll give it a shot.

**Comment by Skaldebane** [comment]
looks good so far!
image

**Comment by CleoMenezesJr** [comment]
> Great! Didn't expect my message to start this chain of events so quickly, but anyways, i'll give it a shot.

The truth is that I was busy with so many other things that I had forgotten. lmao

**Comment by CleoMenezesJr** [comment]
bot, build fm.reaper.Reaper

**Comment by CleoMenezesJr** [comment]
image
